# Homework 6

## Your Name Here (or your names here if you are pair programming)
Riley Sample
rcs3396

## Logistic Regression without using any libraries. 

In [ ]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

# Your allowed to use only the above libraries that are imported. No other libs should be used in this assignment. 

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [ ]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (4 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data, and implement logistic regression model ***without*** using any libs than imported above. You must do it manually.

* Do maximum **100 iterations**
* Use a very small learning rate for checking your GD implementation. 
* Your are allowed to use your choice of learning rate, like using 0.0001, 0.001 or 0.01 or 0.1 or higher/lower. 
* Visualize your costs.
* No need to add an y-intercept in this task. 


In [ ]:
# x and y
y = heart_df[['Target']]
y = y.replace(to_replace="No", value=1)
y = y.replace(to_replace="Yes", value=0)

x = heart_df[['Age', 'Sex', 'Chol']]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

In [ ]:
print(X_train.shape)
print(y_train.shape)

(242, 3)
(242, 1)


In [ ]:
y_train[:10]

Target
102       1
261       0
228       0
288       1
78        1
124       0
200       1
197       1
24        0
174       0

In [ ]:
def sigmoid(z):
    """
    To convert continuous value into a range of 0 to 1

    I/P
    ----------
    z : Continuous value

    O/P
    -------
    Value in range between 0 to 1.
    """
    g = 1 / (1 + np.exp(-z))
    return g

In [ ]:
def cost_function(X, y, weights):
    """
    Compute cost for logistic regression.

    I/P
    ----------
    X : 2D array where each row represent the training example and each column represent the feature ndarray. 
        Dimension (n x d)
        n = number of training examples
        d = number of features (including X_0 column of ones)
    y : 1D array of labels/target value for each traing example. dimension(1 x n)

    weights : 1D array of fitting parameters or weights. Dimension (1 x d)

    O/P
    -------
    cost : The cost of using theta as the parameter for linear regression to fit the data points in X and y.
    """
    n, d = X.shape
    x_dot_weights = X.dot(weights)

    cost = 1.0 / n * (-y.T.dot(np.log(sigmoid(x_dot_weights))) - (1 - y).T.dot(np.log(1 - sigmoid(x_dot_weights))))

    return cost

In [ ]:
def gradient(X, y, weights):
    """
    Compute gradient for logistic regression.

    I/P
    ----------
    X : 2D array where each row represent the training example and each column represent the feature ndarray. 
    Dimension(m x n)
        m = number of training examples
        n = number of features (including X_0 column of ones)
    y : 1D array of labels/target value for each traing example. dimension(1 x m)

    weights : 1D array of fitting parameters or weights. Dimension (1 x n)

    O/P
    -------
    grad: (numpy array)The gradient of the cost with respect to the parameters theta
    """
    n, d = X.shape
    x_dot_weights = X.dot(weights)

    grad = (1.0 / n )* (sigmoid(x_dot_weights) - y).T.dot(X)

    return grad

In [ ]:
weights = np.zeros(3)

cost = cost_function(X_train, y_train, weights)
grad = gradient(X_train, y_train, weights)

print(cost)
print(grad)

Target    0.693147
dtype: float64
        Age  Sex  Chol
0       NaN  NaN   NaN
1       NaN  NaN   NaN
2       NaN  NaN   NaN
3       NaN  NaN   NaN
4       NaN  NaN   NaN
...     ...  ...   ...
298     NaN  NaN   NaN
300     NaN  NaN   NaN
301     NaN  NaN   NaN
302     NaN  NaN   NaN
Target  NaN  NaN   NaN

[243 rows x 3 columns]


In [ ]:
weights = np.zeros(3)


# Now we optimize it using Gradient Descent. 
num_iterations = 100
learnin_rate = 0.01

cost_list = []


# Implementation here is removed. 

# Your task to implement the GD here. 

for i in range(0, num_iterations):
    
    # 
    # Calculate the costs 
    cost = cost_function(X_train, y_train, weights) 
    
    print("Cost is: ", cost)
    # keep the costs for our visualization later 
    
    cost_list.append(cost)
    
    # Calculate the gradients [CODE REMOVED]
    g = gradient(X_train, y_train, weights)
    
    # Use the gradient to update the weights [CODE REMOVED] 
    weights = weights - (learnin_rate*g)
    

Cost is:  Target    0.693147
dtype: float64


ValueError: ignored

In [ ]:
plt.plot(np.arange(num_iterations), cost_list)
plt.show()

# Task 2 - (4 points)

Cacluate the accuracy, Precision, Recall and F1 score of your logistic regression implementaion. 
Print the results. 

You may use equations shown in lecture/slides/examples.

In [ ]:
def predict(weights, X):
    """
    Predict the class between 0 and 1 using learned logistic regression parameters weights.
    Using threshold value 0.5 to convert probability value to class value 

    I/P
    ----------
    X : 2D array where each row represent the training example and each column represent the feature ndarray. 
        Dimension(n x d)
        n= number of training examples
        d= number of features (including X_0 column of ones)

    theta : 1D array of fitting parameters or weights. Dimension (1 x n)

    O/P
    -------
    Class type based on threshold
    """
    p = sigmoid(X.dot(weights)) >= 0.5
    return p.astype(int)

In [ ]:
# Add your code Here! 
from sklearn.metrics import classification_report
pred = predict(weights, X_test)

print(classification_report(y_test, pred))

# Task 3 - (4 points)


Add y-intercept and repeat the above 2 tasks. Do you see any differences after adding the y-intercept? 


In [ ]:
# Add your code Here! 

# Task 4 - Implement the Bold Driver   - (4 points)
In your GD implementation, add the bold driver idea to have a dynamic learning rate.

* Add a stop codition to stop the GD when the cost is not changing more than 0.001. (differences between two costs not more than 0.001, then stop)
* Can you stop earlier than 100 iterations? 




In [ ]:
# Add your code Here! 

# Task 5 - Implement the L2 norm regularization.  - (4 points)

Modify your Cost and gradient to implement the l2 norm regularization. 
Repreat the task 1 and 2 to check if your result is changing. 

* Use y-itercept. 
* Do max 100 iterations as before and report your accuracy, Precision, Recall and F1. 
* You can stop earlier when the cost is not changing than 0.001. 


**Optional:** you might want to use the bold driver. But you can do this task without the bold driver as well. 



In [ ]:
# Add your code Here! 